[RFC] Reconfigure the CI

dave-tucker opened the pull request:
This PR is both an RFC and a test case for reconfiguring the CI.

- Remove existing hacks in place to copy artifacts from ephemeral VM to the LinuxKit CI controller
- Control LTP test runs through `rtf` and labels and not through a make target called by CI directly
- Move the GCP test (run via Python scripts in CI) to a `rtf` test case
- Run the `linuxkit.platforms` tests on PR's and Master from *after* the first test run has been completed

This PR makes the necessary changes in LinuxKit and update the documentation.

Once CI has been update and reconfigured this PR should be green and ready for merge.

justincormack commented:
So whats the new security story for access to cloud resources?

dave-tucker commented:
PR's are tested on an ephemeral VM without an access to cloud resources.
Once tests have passed (which includes running images in QEMU etc...) we should have enough confidence to be able to test these images on GCP and other platforms. This is done from the privileged section of the CI that has access to cloud resources.

dave-tucker commented:
In other words, the security story is unchanged.
The only practical difference is in the source of the image that gets pushed to GCP for testing (which only occurs after a successful test run).

Today, it gets `scp`'d back from the VM.

With these changes, we'll build it ourselves.

rneugeba commented on docs/testing.md:
```diff
@@ -67,28 +67,27 @@ The test logs are available via the `Details` link but also via this [website](h
 This target runs the tests using `rtf` and the results directory is `scp`'ed back to the controller.
-The test results will be stored in DataKit for additional access
-Additionally, the `./artifacts` folder is `scp`'ed back to the controller.
+The test results will be stored in DataKit. 
```
will -> are

rneugeba commented on test/Makefile:
```diff
@@ -1,9 +1,9 @@
-# TODO: should have a separate target
-all: check-deps test-pr ltp
+platforms: check-deps test-platforms
+pr: default
```
what's the difference between `default` and `pr`. Can one of them be removed, or a more common pattern would be to have `default: pr` and then `pr: ...`

dave-tucker replied:
no difference so yes we can remove PR completely.

rneugeba replied:
I wouldn't remove it I would just re-arrange it because it's clearer

dave-tucker replied:
ack

justincormack commented:
Needs a rebase

dave-tucker commented:
Rebased. Addressed Comments. Fixed Flakiness in GCP Tests.

@talex5 here is a summary of the work required in LinuxKitCI:

1) Stop copying the `artifacts` directory - we no longer need these files as we actually have tests that are run during the `make ci-pr`. We do still need the `test/_results` directory

2) **ON PRs ONLY** run `make ci-platforms` from the LinuxKitCI controller (the bit with the necessary creds for talking to GCP) *after* the `make ci-pr` completed successfully in the GCP VM.
This will also generate a `test/_results` directory which will need storing in DataKit.

3) Disable running the GCP tests - this is taken care of in `rtf` and via `make ci-platforms`

4) Disable the LTP tests on master and tag builds - these are now enabled when a `make ci` or `make ci-tag` are done and run by `rtf`

I'm hoping this will be the last set of "major" CI changes for a while.

talex5 commented:
How can we "run `make ci-platforms` from the LinuxKitCI controller" for PRs? The `Makefile` is from the attacker, and the controller has access to the secrets.

dave-tucker commented:
@talex5 I think in this case we should probably whitelist some PR's (e.g those by org owners) and wait for a "ok to test" comment for others. IIRC @justincormack suggested something similar to this in Slack when I last suggested it but the scrollback doesn't go far enough for me to find a link

justincormack commented:
Yes that's an option.

On 22 May 2017 3:08 pm, "Dave Tucker" <notifications@github.com> wrote:

> @talex5 <https://github.com/talex5> I think in this case we should
> probably whitelist some PR's (e.g those by org owners) and wait for a "ok
> to test" comment for others. IIRC @justincormack
> <https://github.com/justincormack> suggested something similar to this in
> Slack when I last suggested it but the scrollback doesn't go far enough for
> me to find a link
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/linuxkit/linuxkit/pull/1853#issuecomment-303110375>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdcPDUyt0aLKzrqGEguOihGA5xEVVOBks5r8Za0gaJpZM4NfOkF>
> .
>

dave-tucker commented:
The changes outlined above will simplify the CI setup and maximise the ability for LinuxKit contributors to make trivial CI changes. CI will call certain `make` targets and `rtf` controls which tests get run through labels etc...

The downside is that in order to test PR's against the cloud platforms we might need to run a potentially compromised `Makefile` on a machine that has access to the cloud platforms we need to test on.

I can see a few options here:

1) Don't worry about it and hope that our tests catch any mischief

2) Don't test running images on the cloud platforms on PR's

3) Have LinuxKitCI detect `Makefile` changes and either fail the build either immediately or at the `Test Cloud Platforms` step. This would then require approval from a maintainer in order to continue the tests.

There are probably other options I've missed so please feel free to make suggestions.

@riyazdf what would you recommend?

justincormack commented:
I think some of the Docker projects eg `notary` require approval to run CI. In this case we could run the base tests but not the platform tests. We would want to be able to whitelist contributors, not just maintainers.

2. is not an option - there is no point in CI if we don't run it.

dave-tucker commented:
@justincormack if we went down the approval route (which I think is best) would you:

a) Run platform tests only if contributor is on the whitelist
b) Run platform tests only if Makefile was changed && contributor is on the whitelist

justincormack commented:
B seems the least friction, but changing Makefile is fairly rare so either
ok

On 23 May 2017 16:39, "Dave Tucker" <notifications@github.com> wrote:

@justincormack <https://github.com/justincormack> if we went down the
approval route (which I think is best) would you:

a) Run platform tests only if contributor is on the whitelist
b) Run platform tests only if Makefile was changed && contributor is on the
whitelist

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/linuxkit/linuxkit/pull/1853#issuecomment-303438999>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAdcPPDUlCuEQqn-YJdyGSqqhGeFZRSVks5r8v3MgaJpZM4NfOkF>
.

dave-tucker commented:
@talex5 it looks like we have a solution for the modification of the `Makefile`.
Would you be able to look at making the changes outlined here in LinuxKitCI?

ijc25 commented:
>  b) Run platform tests only if Makefile was changed && contributor is on the whitelist

Shouldn't that be:
b) Run platform tests only if Makefile was not changed || contributor is on the whitelist
?

Also if we go the `ok for test` route (not sure if we are) then don't we need to protect against the user subsequently pushing more commits to the PR?

And in the "b) Run platform tests only if Makefile was not changed || contributor is on the whitelist" we need to be careful of races between evaluating that condition and pulling from the PR branch, i.e. we should carry the sha which we evaluated over and not repull the branch by name.

talex5 commented:
Has @riyazdf approved this plan? It sounds pretty risky to me, but I'll implement it if he's happy.

dave-tucker commented:
@ijc25 yes you are right. thanks for the correction.

> then don't we need to protect against the user subsequently pushing more commits to the PR?

Yep I think that the approval would be per SHA and not for the PR as a whole.

> And in the "b) Run platform tests only if Makefile was not changed || contributor is on the whitelist" we need to be careful of races between evaluating that condition and pulling from the PR branch, i.e. we should carry the sha which we evaluated over and not repull the branch by name.

Ack. LinuxKitCI deals with SHA's anyway and not branch names so we should be ok in this regard.

dave-tucker commented:
@talex5 ack. waiting on a comment from @riyazdf

riyazdf commented:
I agree with @ijc25: `Run platform tests only if Makefile was not changed || contributor is on the whitelist`.   As @justincormack notes, we have a similar setup in Notary and it's worked well for us.

@dave-tucker: Where are planning on maintaining/storing the whitelist?
@talex5: do you have additional concerns?

Also, instead of `ok to test` can we use the Jenkins terminology `test this please` (or close to it)?  `test this please` is for running one CI run against the current state of the PR, `ok to test` is for running CI for this state and for any future commits.

dave-tucker commented:
@riyazdf the whitelist would be checked in to the `linuxkit/linuxkit-ci` repo.
👍 on all other points

riyazdf commented:
@dave-tucker: 👍  thank you, sounds good!

talex5 commented:
@riyazdf : basically what we want to know is: is it safe to run `make` (a decades old C tool not designed for security) in a directory with a given set of files?

e.g.

- Does `make` read any files besides `Makefile`?
- Does the Makefile include other files (now or in the future)?
- Does the Makefile rule run any software from the checkout (now or in the future)?
- etc

If not, we give all our cloud keys (GCP, AWS, etc) to the attacker.

riyazdf commented:
@talex5: if the Makefile or our setup expands in scope to pull in other files, we'll have to also update the CI gating rules
ex: `Run platform tests only if Makefile && <other_file> was not changed || contributor is on the whitelist

As for `make` itself: I'm ok with using that program - of course we'll have to monitor it for serious security issues.  This is already the case today even if we don't opt to use it in LinuxKit, since we use it across other projects for critical CI operations

WDYT?

ijc25 commented:
Pretty ignorant about our CI setup, but just brainstorming...

> If not, we give all our cloud keys (GCP, AWS, etc) to the attacker.

Do we necessarily need to expose those to all processes spawned by the CI or could we restrict them to just a smaller subset? e.g. a small set of "do the cloud bits" shell scripts to which we apply the restrictions we are talking about adding to modifications of the `Makefile`

Do we have the option of having the CI itself use the primary keys to generate short term, much more restricted keys, either on a per run basis or simply to turn them over more quickly (i.e. daily)?

If not could we run a proxy which accepts short term keys which we control and turns them into requests with the proper keys to the actual cloud services.

Can we use different keys (ephemeral or otherwise) with much more specific sets of abilities for different operations and apply different restrictions based on which ones are more powerful? i.e. read only keys for things which only need that privilege.

Could we put some/all of the bits which need the keys into the CI rather than the repo?

dave-tucker commented:
> Do we necessarily need to expose those to all processes spawned by the CI or could we restrict them to just a smaller subset? e.g. a small set of "do the cloud bits" shell scripts to which we apply the restrictions we are talking about adding to modifications of the Makefile

We're already strict about who has access to the secrets. Alas, "do the cloud bits" isn't a shell script, it's a bunch of tests in `./test/cases` orchestrated by `rtf`.  Having `make` take care of this solves dependency problems (are `rtf`, `moby` and `linuxkit` installed and of the correct version?) and allows us the flexibility to change the labels provided to `rtf` easily.

> Do we have the option of having the CI itself use the primary keys to generate short term, much more restricted keys, either on a per run basis or simply to turn them over more quickly (i.e. daily)?

That's definitely an option. I don't think it'd be much work to generate single use access keys using GCP's IAM API. I'm not sure how that would interface with `docker secrets` though but maybe that's not a concern if the keys are short lived.

> Could we put some/all of the bits which need the keys into the CI rather than the repo?

That's not really an option as it ends up duplicating effort. For example, we've duplicated `linuxkit run gcp` in a set of python scripts in LinuxKitCI instead of dogfooding. That won't scale for all the other cloud platforms we intend on testing.

ijc25 commented:
> Alas, "do the cloud bits" isn't a shell script, it's a bunch of tests in ./test/cases orchestrated by rtf. Having make take care of this solves dependency problems (are rtf, moby and linuxkit installed and of the correct version?) and allows us the flexibility to change the labels provided to rtf easily.

So don't we need to gate changes to the tests in the same way as the `Makefile`, otherwise wouldn't I just write `tests/cases/exfiltrate-secrets.test`? This is essentially @talex5's "Does the Makefile rule run any software from the checkout (now or in the future)?" point I think.

dave-tucker commented:
@ijc25 good point. i'm debating whether or not to suggest that we just gate all PR's that are from non-whitelisted users. that way, we don't run the risk of forgetting to gate certain paths (i.e the tests)

justincormack commented:
@dave-tucker thats probably fine, we will whitelist all the regular contributors pretty quickly. Apparently some of the bits for this have a whitelist from maintainers via comments on the PR which would make it easy.

dave-tucker commented:
ping @talex5 - do you have a response to @riyazdf's comment?
per @ijc25 and @justincormack's comments I think it's sensible to only test PR's from whitelisted contributors unless approval is explicitly given.

where i'd like to eventually take this is per this diagram I pasted in Slack last week:
![linuxkitci flow 1](https://cloud.githubusercontent.com/assets/3781777/26585146/338755f6-4543-11e7-8fb1-08b2cab74012.png)

the tests could be broken down in to a number of containers that can run in parallel.
each container would only contain the secrets required for it's tests (e.g **only gcp**)
ideally, the secrets in the container could be single use and generated by LinuxKitCI itself using the cloud providers IAM APIs (full credit to @ijc25 for that idea)

talex5 commented:
@dave-tucker : As @ijc25 pointed out later, this would require running the test scripts with privileges too, so clearly the `Makefile` isn't the only file that needs to be trusted. Given this, it seems (even more) impossible to secure the system.

And even with whitelisting, you're still effectively giving the credentials to everyone on the whitelist. Who is responsible if a whitelisted contributor misuses them? And if someone submits a buggy PR, who is responsible for finding all the VMs and removing them? The CI can't track them if they're created by the tests.

It's certainly possible to proxy the requests through a restricted API provided by the CI (as @ijc25 suggests), but then you're not testing the tool against the real services.

> For example, we've duplicated linuxkit run gcp in a set of python scripts in LinuxKitCI instead of dogfooding. That won't scale for all the other cloud platforms we intend on testing.

Scaling isn't the problem. We can use `linuxkit run` to test the images on various plaforms; we just have to use a known-good version of it (not the one from the PR). The issue is whether it's worth accepting all the above security risks in order to test the *tool* (in addition to testing the images).

justincormack commented:
@talex5 I trust everyone who is whitelisted, we can add a nightly cron job to test if there are any stray machines created, and yes we absolutely do want to test the tool itself thats the whole point.

riyazdf commented:
@talex5: Maintainers are ultimately responsible for trusting those who they wish to put on the whitelist, and ensuring they will not abuse the system.  Maintainers are collectively responsible for any security response, CI related or otherwise.

I'm in agreement with @justincormack that I trust the folks on the whitelist and additional cron jobs and tests will help keep us on top of things.  We absolutely need to test the tool in an automated fashion across multiple platforms, so it's great that we're having this in-depth discussion to ensure we do so in a way that is safe.

rneugeba commented:
Just to chime in, let's not try to do any gymnastics trying to figure out if a PR is safe to run by an untrusted, ie not-whitelisted, user. Nothing should be run unless the PR is from a whitelisted user or a whitelisted user (or maintainer) hits a button on the PR. That way we can keep the whitelist in linuxkit-ci but move more of the logic of what is run on PRs/merge/etc in linuxkit's Makefiles

ijc25 commented:
Perhaps we should arrange for something (CI, gordon, poule, etc) to label PRs which need checking by a maintainer and marking as safe, so we can filter on them?

dave-tucker commented:
> That way we can keep the whitelist in linuxkit-ci but move more of the logic of what is run on PRs/merge/etc in linuxkit's Makefiles

It did occur to me that it might be easier for LinuxKitCI to be configured using a YAML file similar to that of CircleCI. Each section (e.g `checkout`, `build`. `test`, `deploy`) would be an array of commands that would be executed in order. Defaults for each section would be documented. They could be overridden in a `linuxkitci.yml` which could be stored either in this repo or in the LinuxKitCI repo.

This would also document the relationship between LinuxKitCI and Makefiles so there was less magic involved.

justincormack commented:
The whole whitelisting thing is well understood and we should not try to reinvent anything...

justincormack commented:
eg https://github.com/jenkinsci/ghprb-plugin is the canonical example

talex5 commented:
Given that we're not actually using any features of DataKitCI (the "pipeline" is a single step that calls `make`, which then coordinates everything), perhaps we should just use Jenkins?

dave-tucker commented:
@talex5 while that may be true as of now, I'd very much like for it to be more than one `make` incantation and would be happy to work with you to make better use of DatakitCI's features. I think the diagram I linked above goes a good way to explaning what I have in mind - comments welcome. I think we also all agree that having logs in git is much better that anything Jenkins provides today.

It seems we have consensus that testing PR's from only from whitelisted contributors is a reasonable way to protect our secrets while also enabling the `linuxkit` tool to be tested.

We also have some existing whitelisting implementations to borrow from.

The question is whether or not this feature can be added to DataKitCI and how soon we could make use of it.

justincormack commented:
Thinking about where to store the whitelist, I think having it in a file in the CI repo would be fine, happy to do PRs to add people there. Else it could be in the repo itself but it would have to be in the master branch not the checked out branch...

talex5 commented:
@samoht: can we get the PR owner from the bridge metadata?

samoht commented:
@talex5 yup sure, I am adding this shortly. Let me know if you need more metadata.

dave-tucker commented:
Closing as I'm not actively working on this.
